File: CICD Agent/runtime/core/memory_store.py
"""Lightweight per-repo memory store.

Backed by the same SQLite database as the indexer (see schema.sql). No
secrets are stored: tokens, passwords and API keys belong in the OS keyring
and/or env vars.
"""

from __future__ import annotations

import fnmatch
import json
import time
from dataclasses import dataclass
from pathlib import Path
from typing import Any

from runtime.index.database import connect


@dataclass
class PRHistoryEntry:
    id: int
    task_id: str
    pr_id: int | None
    pr_url: str
    title: str
    summary: str
    risk_level: str
    created_at: int


class MemoryStore:
    def __init__(self, repo_path: Path):
        self.repo_path = repo_path.resolve()
        self.conn = connect(self.repo_path)

    def close(self) -> None:
        try:
            self.conn.close()
        except Exception:
            pass

    # --- repo profile (free-form key/value) ---

    def set_profile(self, key: str, value: Any) -> None:
        self.conn.execute(
            "INSERT INTO repo_profile(key, value, updated_at) VALUES(?, ?, ?) "
            "ON CONFLICT(key) DO UPDATE SET value=excluded.value, updated_at=excluded.updated_at",
            (key, json.dumps(value), int(time.time())),
        )

    def get_profile(self, key: str, default: Any = None) -> Any:
        row = self.conn.execute(
            "SELECT value FROM repo_profile WHERE key = ?", (key,)
        ).fetchone()
        if not row:
            return default
        try:
            return json.loads(row["value"])
        except Exception:
            return default

    # --- PR history ---

    def record_pr(
        self,
        *,
        task_id: str,
        pr_id: int | None,
        pr_url: str,
        title: str,
        summary: str,
        risk_level: str = "low",
    ) -> int:
        cur = self.conn.execute(
            "INSERT INTO pr_history(task_id, pr_id, pr_url, title, summary, risk_level, "
            "created_at) VALUES (?, ?, ?, ?, ?, ?, ?)",
            (
                task_id,
                pr_id,
                pr_url,
                title,
                summary,
                risk_level,
                int(time.time()),
            ),
        )
        return int(cur.lastrowid or 0)

    def recent_prs(self, limit: int = 20) -> list[PRHistoryEntry]:
        rows = self.conn.execute(
            "SELECT id, task_id, pr_id, pr_url, title, summary, risk_level, created_at "
            "FROM pr_history ORDER BY id DESC LIMIT ?",
            (limit,),
        ).fetchall()
        return [
            PRHistoryEntry(
                id=int(r["id"]),
                task_id=str(r["task_id"]),
                pr_id=int(r["pr_id"]) if r["pr_id"] is not None else None,
                pr_url=str(r["pr_url"]),
                title=str(r["title"]),
                summary=str(r["summary"]),
                risk_level=str(r["risk_level"]),
                created_at=int(r["created_at"]),
            )
            for r in rows
        ]

    # --- reviewer mapping ---

    def set_reviewer(self, path_glob: str, reviewers: list[str]) -> None:
        self.conn.execute(
            "INSERT INTO reviewer_map(path_glob, reviewers) VALUES(?, ?) "
            "ON CONFLICT(path_glob) DO UPDATE SET reviewers=excluded.reviewers",
            (path_glob, ",".join(reviewers)),
        )

    def reviewers_for_paths(self, paths: list[str]) -> list[str]:
        rows = self.conn.execute(
            "SELECT path_glob, reviewers FROM reviewer_map"
        ).fetchall()
        out: set[str] = set()
        for r in rows:
            pat = str(r["path_glob"]).replace("\\", "/")
            for p in paths:
                if fnmatch.fnmatch(p.replace("\\", "/"), pat):
                    for rv in str(r["reviewers"]).split(","):
                        rv = rv.strip()
                        if rv:
                            out.add(rv)
                    break
        return sorted(out)

    # --- conventions ---

    def add_convention(self, scope: str, rule: str) -> int:
        cur = self.conn.execute(
            "INSERT INTO conventions(scope, rule) VALUES (?, ?)", (scope, rule)
        )
        return int(cur.lastrowid or 0)

    def all_conventions(self) -> list[tuple[str, str]]:
        rows = self.conn.execute(
            "SELECT scope, rule FROM conventions ORDER BY id"
        ).fetchall()
        return [(str(r["scope"]), str(r["rule"])) for r in rows]

    # --- flaky tests ---

    def mark_flaky(self, test_id: str, notes: str = "") -> None:
        self.conn.execute(
            "INSERT INTO known_flaky_tests(test_id, last_seen, notes) VALUES(?, ?, ?) "
            "ON CONFLICT(test_id) DO UPDATE SET last_seen=excluded.last_seen, "
            "notes=excluded.notes",
            (test_id, int(time.time()), notes),
        )

    def is_flaky(self, test_id: str) -> bool:
        row = self.conn.execute(
            "SELECT 1 FROM known_flaky_tests WHERE test_id = ?", (test_id,)
        ).fetchone()
        return row is not None

    def known_flaky_tests(self) -> list[str]:
        rows = self.conn.execute(
            "SELECT test_id FROM known_flaky_tests ORDER BY test_id"
        ).fetchall()
        return [str(r["test_id"]) for r in rows]

    # --- ignored paths ---

    def add_ignored_path(self, path_glob: str) -> None:
        self.conn.execute(
            "INSERT OR IGNORE INTO ignored_paths(path_glob) VALUES (?)", (path_glob,)
        )

    def ignored_paths(self) -> list[str]:
        rows = self.conn.execute("SELECT path_glob FROM ignored_paths").fetchall()
        return [str(r["path_glob"]) for r in rows]

File: packages/core/src/memoryStore.ts
import type { Database as DbType } from "better-sqlite3";
import { openRepoDb, type RepoDatabase } from "./db/database.js";

export interface PRHistoryEntry {
  id: number;
  taskId: string;
  prId: number | null;
  prUrl: string;
  title: string;
  summary: string;
  riskLevel: string;
  createdAt: number;
}

function globToRegex(glob: string): RegExp {
  const escaped = glob.replace(/[.+^${}()|[\]\\]/g, "\\$&").replace(/\\\\/g, "/");
  const re = escaped.replace(/\*\*/g, ".*").replace(/\*/g, "[^/]*").replace(/\?/g, ".");
  return new RegExp("^" + re + "$");
}

export class MemoryStore {
  private readonly handle: RepoDatabase;
  private readonly db: DbType;

  constructor(repoPath: string) {
    this.handle = openRepoDb(repoPath);
    this.db = this.handle.db;
  }

  close(): void {
    this.handle.close();
  }

  setProfile(key: string, value: unknown): void {
    const now = Math.floor(Date.now() / 1000);
    this.db
      .prepare(
        "INSERT INTO repo_profile(key, value, updated_at) VALUES(?, ?, ?) " +
          "ON CONFLICT(key) DO UPDATE SET value=excluded.value, updated_at=excluded.updated_at",
      )
      .run(key, JSON.stringify(value), now);
  }

  getProfile<T = unknown>(key: string, defaultValue: T | null = null): T | null {
    const row = this.db
      .prepare("SELECT value FROM repo_profile WHERE key = ?")
      .get(key) as { value: string } | undefined;
    if (!row) return defaultValue;
    try {
      return JSON.parse(row.value) as T;
    } catch {
      return defaultValue;
    }
  }

  recordPr(args: {
    taskId: string;
    prId: number | null;
    prUrl: string;
    title: string;
    summary: string;
    riskLevel?: string;
  }): number {
    const now = Math.floor(Date.now() / 1000);
    const r = this.db
      .prepare(
        "INSERT INTO pr_history(task_id, pr_id, pr_url, title, summary, risk_level, " +
          "created_at) VALUES (?, ?, ?, ?, ?, ?, ?)",
      )
      .run(
        args.taskId,
        args.prId,
        args.prUrl,
        args.title,
        args.summary,
        args.riskLevel ?? "low",
        now,
      );
    return Number(r.lastInsertRowid);
  }

  recentPrs(limit = 20): PRHistoryEntry[] {
    const rows = this.db
      .prepare(
        "SELECT id, task_id, pr_id, pr_url, title, summary, risk_level, created_at " +
          "FROM pr_history ORDER BY id DESC LIMIT ?",
      )
      .all(limit) as Array<{
      id: number;
      task_id: string;
      pr_id: number | null;
      pr_url: string;
      title: string;
      summary: string;
      risk_level: string;
      created_at: number;
    }>;
    return rows.map((r) => ({
      id: r.id,
      taskId: r.task_id,
      prId: r.pr_id,
      prUrl: r.pr_url,
      title: r.title,
      summary: r.summary,
      riskLevel: r.risk_level,
      createdAt: r.created_at,
    }));
  }

  setReviewer(pathGlob: string, reviewers: string[]): void {
    this.db
      .prepare(
        "INSERT INTO reviewer_map(path_glob, reviewers) VALUES(?, ?) " +
          "ON CONFLICT(path_glob) DO UPDATE SET reviewers=excluded.reviewers",
      )
      .run(pathGlob, reviewers.join(","));
  }

  reviewersForPaths(paths: string[]): string[] {
    const rows = this.db
      .prepare("SELECT path_glob, reviewers FROM reviewer_map")
      .all() as Array<{ path_glob: string; reviewers: string }>;
    const out = new Set<string>();
    for (const r of rows) {
      const re = globToRegex(r.path_glob);
      const matches = paths.some((p) => re.test(p.replace(/\\/g, "/")));
      if (matches) {
        for (const rv of r.reviewers.split(",")) {
          const trimmed = rv.trim();
          if (trimmed) out.add(trimmed);
        }
      }
    }
    return [...out].sort();
  }

  addConvention(scope: string, rule: string): number {
    const r = this.db
      .prepare("INSERT INTO conventions(scope, rule) VALUES (?, ?)")
      .run(scope, rule);
    return Number(r.lastInsertRowid);
  }

  allConventions(): Array<{ scope: string; rule: string }> {
    return this.db
      .prepare("SELECT scope, rule FROM conventions ORDER BY id")
      .all() as Array<{ scope: string; rule: string }>;
  }

  markFlaky(testId: string, notes = ""): void {
    const now = Math.floor(Date.now() / 1000);
    this.db
      .prepare(
        "INSERT INTO known_flaky_tests(test_id, last_seen, notes) VALUES(?, ?, ?) " +
          "ON CONFLICT(test_id) DO UPDATE SET last_seen=excluded.last_seen, notes=excluded.notes",
      )
      .run(testId, now, notes);
  }

  isFlaky(testId: string): boolean {
    const row = this.db
      .prepare("SELECT 1 AS one FROM known_flaky_tests WHERE test_id = ?")
      .get(testId);
    return Boolean(row);
  }

  knownFlakyTests(): string[] {
    const rows = this.db
      .prepare("SELECT test_id FROM known_flaky_tests ORDER BY test_id")
      .all() as Array<{ test_id: string }>;
    return rows.map((r) => r.test_id);
  }

  addIgnoredPath(pathGlob: string): void {
    this.db.prepare("INSERT OR IGNORE INTO ignored_paths(path_glob) VALUES (?)").run(pathGlob);
  }

  ignoredPaths(): string[] {
    const rows = this.db.prepare("SELECT path_glob FROM ignored_paths").all() as Array<{
      path_glob: string;
    }>;
    return rows.map((r) => r.path_glob);
  }
}
